### ISSUE TITLE
Nondeterministic failures (Julia crashes) on Base Julia CI on `tester_win64`

### ISSUE THREAD

**Issue by DilumAluthge** [JuliaSparse/SuiteSparse.jl#43]
Example log: https://build.julialang.org/#/builders/65/builds/4081

**Comment by Gnimuc** [comment]
I'm not sure. This commit might be related: https://github.com/JuliaLang/SuiteSparse.jl/commit/bb068bbd9848a47e6527336231ded17a9cd1951d. But it should be fixed.

Is SuiteSparse.jl the only user of `detect_ambiguities`? I guess no. To me, this looks like the `detect_ambiguities` is broken.

**Comment by Gnimuc** [comment]
This is probably caused by some compiler internal changes related to inference.

**Comment by DilumAluthge** [comment]
That line just lists all of the test sets that Buildbot ran. E.g. Buildbot ran the `ambiguous` test set, the `compiler/inference` test set, etc.

If you scroll up in the log, you can see which test sets passed and which test sets failed.

All of the test sets are passing except for SuiteSparse. For example, the `ambiguous` test set is passing, the `compiler/inference` test set is passing, etc.

**Comment by DilumAluthge** [comment]
It seems relatively rare. You may need to run it for a long time.

**Comment by DilumAluthge** [comment]
This doesn't have anything to do with `detect_amviguities`.

**Comment by Gnimuc** [comment]
OK, another question. If `Base.runtests(["SuiteSparse"])` could trigger a crash, how could I take a snapshot of the current stack for debugging later. Does rr support Windows now?

**Comment by ViralBShah** [comment]
No

**Comment by ViralBShah** [comment]
The crash happens in the first suitsparse test which is detect_ambiguities.

**Comment by DilumAluthge** [comment]
> The crash happens in the first suitsparse test which is detect_ambiguities.

How do you know the crash is happening in the first SuiteSparse test?

**Comment by DilumAluthge** [comment]
> The crash happens in the first suitsparse test which is detect_ambiguities.

How do you know the crash is happening in the first SuiteSparse test?

**Comment by Gnimuc** [comment]
Looks like [the first faliure] is triggered by https://github.com/JuliaLang/julia/commit/a512f1a00fec3bbaa96f83f7eabfdfcba739e587.

**Comment by ViralBShah** [comment]
It shows only 1 error and then crashes (in the tests table). So guessing it is the first or second test.

**Comment by ViralBShah** [comment]
> Looks like [the first faliure](https://build.julialang.org/#/builders/65/builds/3866) is triggered by [JuliaLang/julia@a512f1a](https://github.com/JuliaLang/julia/commit/a512f1a00fec3bbaa96f83f7eabfdfcba739e587).
> 
> Could this `OPENBLAS_MAIN_FREE` env variable affect suitesparse or its testset?

I doubt it.

**Comment by DilumAluthge** [comment]
> It shows only 1 error and then crashes (in the tests table). So guessing it is the first or second test.

I think that might be misleading though.

The tests are run in a separate worker process. Once the worker process has run all of the tests, it reports the results (successes, failures, errors, and broken) back to the main process. At that point, the main process prints the test results to stdout.

The key point here (if I understand correctly) is that the successes/failures/errors/broken information is not communicated back to the main process until the worker process has finished running all of the tests. Therefore, if the worker process crashes before it has finished running all of the tests, all of the successes/failures/errors/broken information will be lost, and therefore the main process will not have any of that information. As a result, the main process only reports that there was "one error" during the SuiteSparse test suite. In reality, we don't know how many tests were run before the worker process crashed.

This is my understanding of it. It would be good if one of the experts (@staticfloat @Keno @vtjnash) can confirm that my understanding and explanation are correct.

**Comment by Gnimuc** [comment]
Is there a way to search a build with its number in Buildbot?

**Comment by Wimmerer** [comment]
@Gnimuc The only change there is moving from a manual function -> Clang generated one. The signatures _do_ differ slightly (a `Ptr{Cvoid}` in the old version while Clang chooses a `Ptr{Ptr{Cvoid}}`, but I doubt that's the issue? This is only happening on the threading test right?

**Comment by Gnimuc** [comment]
According to the log, the error occurs in `test\umfpack.jl:22`. 

> The signatures do differ slightly (a Ptr{Cvoid} in the old version while Clang chooses a Ptr{Ptr{Cvoid}}, but I doubt that's the issue?

[This example](https://github.com/PetterS/SuiteSparse/blob/master/UMFPACK/Demo/umfpack_di_demo.c#L144) shows `Ptr{Ptr{Cvoid}}` is correct. 

BTW, the `tmp` variable `Vector{Ptr{Cvoid}}(undef, 1)` looks weird to me. I would rather use `Ref{Ptr{Cvoid}}(C_NULL)` instead.

**Comment by Wimmerer** [comment]
I don't think I modified that line, but once we've solved this issue I can go back and clean some things up.

UMFPACK and SPQR, the two solvers where we're seeing errors (this one for UMFPACK, the KLU wrapper is blocked by a similar Windows x86_64 only error for threading SPQR), do use BLAS calls, and both now have errors that _shouldn't_ be related to any recent changes.

**Comment by Gnimuc** [comment]
For future reference, 

| Win64 Failures | Caused by SuiteSparse 
|:-:|:-:|
| 3866 | ☑️
| 3863 | ☑️
| 3862 | ❎
| 3861 | ❎
| 3849 | ❎
| 3845 | ❎
| 3844 | ❎
| 3841 | ❎
| 3831 | ❎
| 3826 | ❎
| 3815 | ☑️
| 3814 | ☑️
| 3808 | ❎
| 3797 | ❎
| 3788 | ❎
| 3780 | ❎
| 3779 | ❎
| 3778 | ❎
| 3777 | ❎
| 3774 | ❎
| 3772 | ❎
| 3771 | ☑️
| 3767 | ❎  
| 3765 | ☑️  
| 3764 | ❎  
| 3761 | ❎  
| 3760 | ❎  
| 3757 | ❎  
| 3749 | ❎   
| 3748 | ❎ 
| 3743 | ❎   
| 3734 | ❎   
| 3726 | ❎ 
| 3724 | ❎          
| 3723 | ☑️            
| 3717 | ❎
| 3713 | ❎
| 3711 | ☑️
| 3706 | ❎ 
| 3699 | ☑️
| 3697 | ❎ 
| 3686 | ☑️ 
| 3677 | ❎
| 3675 | ❎
| 3670 | ❎
| 3668 | ❎
| 3665 | ❎
| 3660 | ❎
| 3659 | ❎
| 3644 | ❎
| 3640 | ❎
| 3635 | ❎ 
| 3631 | ❎ 
| 3630 | ❎ 
| 3625 | ❎ 
| 3622 | ❎
| 3615 | ❎

**Comment by Gnimuc** [comment]
> Note. This test is failing for `Int32` indices (this would naturally be the first failure, so that's the likely reason). I suppose it's possible that we're using an Int64 (long) `numeric` struct where we shouldn't be, but that should error on other platforms as well.

@Wimmerer  Could these lines be the reason? 

https://github.com/JuliaLang/SuiteSparse.jl/blob/b15c39be53f7823c721c1f8a7c036105e2baa04a/src/LibSuiteSparse.jl#L8-L13

**Comment by Wimmerer** [comment]
I'm not familiar enough with the internals of SuiteSparse to identify that for sure, but it's a very probable culprit given that's our only failing plat. 

I'm not clear on what that's doing though. It's used to set `SuiteSparse_long` and `SuiteSparse_long_max` to the correct values. But we're not failing when it should be using `SuiteSparse_long`, which should be used when we call `dl_numeric` not `di_numeric`.

**Comment by Wimmerer** [comment]
Looking into it, that _should_ be fine. And for this failing test it shouldn't matter, we should be using an Int32 numeric struct. Let me test that is actually happening

**Comment by Gnimuc** [comment]
`SuiteSparse_long` is set to `long long` in the BB2 script: 

https://github.com/JuliaPackaging/Yggdrasil/blob/a491caf99504e08f9a9ee3878a54c59c0bb81d45/S/SuiteSparse/SuiteSparse/build_tarballs.jl#L30  

so we do the same thing on the Julia side.

**Comment by Gnimuc** [comment]
And, I can not reproduce the nondeterministic failure on my machine. I've run like 200 times without a single crash. But the rate on CI is roughly 1 per 25 times on average.

**Comment by Wimmerer** [comment]
This is why I didn't use a global constant. Each UMFPACKLU should have its own `control` field.

**Comment by Wimmerer** [comment]
I will check with him. There are 2 different issues, that seem to have occurred at a similar time going on here:
1. UMFPackLU fails on Windows x86_64 on the `umfpack_di_numeric` call. This occurs in single-threaded usage. That is this issue.
2. Tests from the KLU PR fail on the unrelated threading test of SPQR on Windows x86_64

Both of these occur only on CI, and only on x86_64.

@ViralBShah I have confirmed with Dr. Davis that everything should be thread safe. We'll probably need to remove the global from umfpack.jl which I can do in a bit. But both of the above tests have worked perfectly fine until some sort of changes to nightly IMHO.

**Comment by DilumAluthge** [comment]
Here's another log that has the extra details: https://build.julialang.org/#/builders/65/builds/4256

**Comment by ViralBShah** [comment]
The updated openblas PR was merged after the latest failure that @DilumAluthge posted. Let's see if more failures show up. This is literally failing in the very first test - and this makes me hopeful that there was no real issue with suitesparse.

**Comment by ViralBShah** [comment]
Still happening: https://build.julialang.org/#builders/65/builds/4280

**Comment by ViralBShah** [comment]
The comments in here discuss how SuiteSparse always using 64-bit ints on all 64-bit platforms including Win64.
 https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/SuiteSparse_config/SuiteSparse_config.h

**Comment by Gnimuc** [comment]
I'm confused that why this only happened nondeterministically on CI. If the typesize is wrong, shouldn't it fail every time?

**Comment by ViralBShah** [comment]
Given that the failure was in `umfpack_di_*` routine, perhaps the LONG_MAX fix above might help.

https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/00d3d9387671a02d5e8ea8753faa8db050ae276b/UMFPACK/Source/umf_internal.h#L254

```
/*
    NUMERIC_VALID and SYMBOLIC_VALID:
    The different values of SYBOLIC_VALID and NUMERIC_VALID are chosen as a
    first defense against corrupted *Symbolic or *Numeric pointers passed to an
    UMFPACK routine.  They also ensure that the objects are used only by the
    same version that created them (umfpack_di_*, umfpack_dl_*, umfpack_zi_*,
    or umfpack_zl_*).  The values have also been changed since prior releases of
    the code to ensure that all routines that operate on the objects are of the
    same release.  The values themselves are purely arbitrary.  The are less
    than the ANSI C required minimums of INT_MAX and LONG_MAX, respectively.
*/

```

**Comment by Gnimuc** [comment]
I think we should find the first failure case and we may get clues from those related commits.

I've manually checked those builds after 3615. The results are listed here: 

https://github.com/JuliaLang/SuiteSparse.jl/issues/43#issuecomment-944899466

The first case is probably 3686, but I'm not sure, I failed to retrieve those build logs before 3615.

**Comment by ViralBShah** [comment]
@DilumAluthge Are you still seeing these crashes after bumping SuiteSparse.jl on master?

**Comment by DilumAluthge** [comment]
Not sure. I'll keep an eye out. If we go a while without seeing them, we can close this.

**Comment by ViralBShah** [comment]
@DilumAluthge Closing this for now. Let's reopen if we find it happening again.

**Comment by moyner** [comment]
I'm seeing nondeterministic triggering of a similar crash in a local simulation code on Windows. This seems to happen during the first (precompile) solve when -O3 is enabled. The inputs to this simulation are static and do not contain any randomization. I'm attaching the crash log (where I have replaced my homedir with MYPATH). Windows 10 64 bits on Intel, Julia 1.7.2 official binary.

Unfortunately the simulation data is confidential but if there is interest I can try to find a non-confidential variant that also crashes. I could also make a separate issue if this is believed to be a different issue.

[crash.txt](https://github.com/JuliaSparse/SuiteSparse.jl/files/8107728/crash.txt)
.

**Comment by ViralBShah** [comment]
Is it possible to reliably reproduce the failure? And can you reproduce it with the default level of optimization?

**Comment by mkitti** [comment]
> This seems to happen during the first (precompile) solve when -O3 is enabled

If you do not use `-O3`, do you still encounter the issue?

**Comment by Wimmerer** [comment]
In `crash.txt` I don't see the file `x86_64-w64-mingw32.jl` like I would expect. Line 388, which is the last Julia line in `crash.txt` only makes sense before commit https://github.com/JuliaSparse/SuiteSparse.jl/commit/b15c39be53f7823c721c1f8a7c036105e2baa04a. 

Indeed in 1.7 we don't have that commit. Does that need a backport?

**Comment by Wimmerer** [comment]
I believe this was an issue with signatures? The difference seems to be in the `Numeric` argument. I'm not sure _why_ this is nondeterministic though? 

Original:

```julia
tmp = Vector{Ptr{Cvoid}}(undef, 1)
status = ccall(($num_r, :libumfpack), $itype,
                           (Ptr{$itype}, Ptr{$itype}, Ptr{Float64}, Ptr{Cvoid}, Ptr{Cvoid},
                            Ptr{Float64}, Ptr{Float64}),
                           U.colptr, U.rowval, U.nzval, U.symbolic, tmp,
                           umf_ctrl, umf_info)
```
New:

```julia
function umfpack_dl_numeric(Ap, Ai, Ax, Symbolic, Numeric, Control, Info)
    @ccall libumfpack.umfpack_dl_numeric(Ap::Ptr{Clonglong}, Ai::Ptr{Clonglong}, Ax::Ptr{Cdouble},
Symbolic::Ptr{Cvoid}, Numeric::Ptr{Ptr{Cvoid}}, Control::Ptr{Cdouble}, Info::Ptr{Cdouble})::Clonglong
end
tmp = Vector{Ptr{Cvoid}}(undef, 1)
status = $num_r(U.colptr, U.rowval, U.nzval, U.symbolic, tmp, umf_ctrl, umf_info)
```

**Comment by Gnimuc** [comment]
Converting VectorPtr to Ptr calls different Base.unsafe_convert subroutines than converting to PtrPtr. I'm on a phone, so can't add the link to the code here, but you can do a quick check using at-which in Julia REPL.

As gc rooting is running after llvm optimization, my guess is that after running some aggressive optimization pass, the Vector is optimized out before gc rooting because of the wrong signature(if the signature is correct, ccall will add it to an extra argument list of the foreigncall during lowering, which prevents it from being optimized out)

**Comment by Wimmerer** [comment]
```julia
unsafe_convert(::Type{Ptr{T}}, a::Array{T}) where {T} = ccall(:jl_array_ptr, Ptr{T}, (Any,), a)
unsafe_convert(::Type{Ptr{S}}, a::AbstractArray{T}) where {S,T} = convert(Ptr{S}, unsafe_convert(Ptr{T}, a))
```
I believe the first form is what's being done in the new code, because the parameter types are the same. The second is what's probably being done in the old code.

@moyner could you maybe try building Julia from master or https://github.com/JuliaLang/julia/tree/release-1.8-beta1 and running your experiment? It should be fixed by the new ccall signature.

**Comment by ViralBShah** [comment]
I don't think anyone has complained on 1.6 - so best not to backport anything to 1.6 until we have a confirmed case there or a reproducible issue for 1.6.

**Comment by ViralBShah** [comment]
I am able to reproduce this issue on my Win64 VM reliably. It happens only when the Symbolic LU factorization reuse tests are included - so that test seems to be corrupting something. Either the test, or the code implementing that part is the possible culprit.

**Comment by ViralBShah** [comment]
Looking further into those tests, it seems odd to be testing for all these types: `Float16, Float32, ComplexF16, ComplexF32`, which clearly we do not support with UMFPACK. It is surprising that it only fails on Windows. Removing those data types fixes the test failure.

**Comment by Wimmerer** [comment]
Okay, I wonder why I wasn't hitting the failures on my machine. Perhaps I had the wrong branch checked out.

I will try to see what is causing those failures, since indeed it would be bad for users to hit this from a failure to promote.

**Comment by ViralBShah** [comment]
This is the smallest test example I can get to fail. However, it doesn't fail when running interactively. It has to be run through the tests.

```
using Test
using SuiteSparse.UMFPACK
using SuiteSparse
using SuiteSparse: increment!
using LinearAlgebra:
    I, det, issuccess, ldiv!, lu, lu!, Adjoint, Transpose, SingularException, Diagonal
using SparseArrays: nnz, sparse, sprand, sprandn, SparseMatrixCSC

    A0 = sparse(increment!([0,4,1,1,2,2,0,1,2,3,4,4]),
                increment!([0,4,0,2,1,2,1,4,3,2,1,2]),
                [2.,1.,3.,4.,-1.,-3.,3.,6.,2.,1.,4.,2.], 5, 5)
    A1 = sparse(increment!([0,4,1,1,2,2,0,1,2,3,4,4]),
               increment!([0,4,0,2,1,2,1,4,3,2,1,2]),
               [2.,1.,3.,4.,-1.,-3.,3.,9.,2.,1.,4.,2.], 5, 5)
   Tv = Float16
   Ti = Int32
   A = convert(SparseMatrixCSC{Tv,Ti}, A0)
   B = convert(SparseMatrixCSC{Tv,Ti}, A1)
   b = Tv[8., 45., -3., 3., 19.]
   F = lu(A)
   lu!(F, B)
  @test F\b ≈ B\b ≈ Matrix(B)\b
```

**Comment by Wimmerer** [comment]
> It has to be run through the tests.

Why is this do we know? Compiler optimizing things out? 

Okay I'll investigate and make a PR. Should I make a separate PR to remove those tests for other eltypes? Or should we keep them and ensure they promote correctly.

**Comment by ViralBShah** [comment]
Well, this doesn't seem related to SuiteSparse at all. Seems like a Julia bug on Win64. This crashes:

```
x = rand(Float16, 5)
y = rand(Float16, 5)
x ≈ y
```

**Comment by Wimmerer** [comment]
```
From worker 2:	umfpack_di_numeric at C:\buildbot\worker-tabularasa\tester_win64\build\bin\libumfpack.DLL (unknown line)
      From worker 2:	umfpack_di_numeric at C:\buildbot\worker-tabularasa\tester_win64\build\share\julia\stdlib\v1.9\SuiteSparse\lib\x86_64-w64-mingw32.jl:1748
```

Separate bugs? Or related? That seems like it's happening inside SuiteSparse to me (that's from the Julia build https://build.julialang.org/#/builders/63/builds/6099/steps/5/logs/stdio)

**Comment by ViralBShah** [comment]
What I found is that I can't trigger the bug unless I include the test that does norm of the Float16 array. That makes me believe that it corrupts something during codegen which makes the first thing to run in the SuiteSparse tests fail.

**Comment by ViralBShah** [comment]
We'll have to temporarily disable these tests so that everything passes and we can go ahead with bumping SparseArrays and removing from sysimage and so on.
